src/rng.rs

```diff
@@ -1,80 +1,778 @@
 use crate::Vec;
-use ark_ff::{FromBytes, ToBytes};
-use ark_std::convert::From;
-use ark_std::marker::PhantomData;
-use ark_std::rand::{RngCore, SeedableRng};
-use digest::Digest;
-
-/// An RNG suitable for Fiat-Shamir transforms
-pub trait FiatShamirRng: RngCore {
-    /// Create a new `Self` with an initial input
-    fn initialize<'a, T: 'a + ToBytes>(initial_input: &'a T) -> Self;
-    /// Absorb new inputs into state
-    fn absorb<'a, T: 'a + ToBytes>(&mut self, new_input: &'a T);
-}
+use ark_crypto_primitives::sponge::poseidon::{PoseidonConfig, PoseidonSponge};
+use ark_crypto_primitives::sponge::{Absorb, CryptographicSponge};
+use ark_ff::PrimeField;
+
+use ark_std::rand::RngCore;
 
 /// A simple `FiatShamirRng` that refreshes its seed by hashing together the previous seed
 /// and the new seed material.
-pub struct SimpleHashFiatShamirRng<D: Digest, R: RngCore + SeedableRng> {
-    r: R,
-    seed: [u8; 32],
-    #[doc(hidden)]
-    digest: PhantomData<D>,
-}
+/// Exposes a particular instantiation of the Poseidon sponge
 
-impl<D: Digest, R: RngCore + SeedableRng> RngCore for SimpleHashFiatShamirRng<D, R> {
+#[derive(Clone)]
+pub struct SimplePoseidonRng<F: PrimeField>(PoseidonSponge<F>);
+
+impl<F: PrimeField> RngCore for SimplePoseidonRng<F> {
```

Comment:
I think instead of implementing `RngCore` for `SimplePoseidonRng`, we should just directly squeeze out challenges. That is, we should change the verifier and prover to invoke methods from `CryptographicSponge`, as opposed to methods from `RngCore`. The reason for this is two-fold. 

First, absorptions can be more efficient: we can absorb elements that are representable as field elements directly. 

Second, challenge generation is more efficient: the `RngCore` impl first invokes Poseidon to generate field elements, which it convert to bits to feed into the Rng, and then the `Field::rand` impl uses these bits to sample a field element. So we're doing a useless round trip here.

Reply:
RngCore is required for backwards compatibility with generation of random field elements in ark-ff and ark-poly. Changing these is a much larger piece of work than exposing the sponge construction through RngCore and beyond the scope of this PR imo.